### PR TITLE
fix #18149 - insert measure after selection (don't move keysig, timesig and clef)

### LINKED ISSUE #18149: Append measures after a measure before a meter change should keep old meter
### Your idea

The command to append measures after a selection will use the time signature of the *following* measure if different from the time signature of the selected measure.  This is coiunter-intuitive and also misses out on an opportunity to solve a long-standing problem.

### Problem to be solved

MU3 only had a command to insert before a selected a measure, not to append after.  There was a never-ending debate about which time signature should be used if you tried to insert the first measure of a new time signature, with either answer being right half the time and wrong half the time.

Now that we have an append after command as well, it would be great if it's behavior were actually different from insert before.  So in this example:

image

Selecting the 4/4 bar and choosing append after would give you new 4/4 bars, while selecting the 3/4 bar and choosing insert before would give you new 3/4 bars.  Because actually creating more 4/4 bars in that scenario is unnecessarily difficult (you need to insert *before* the existing 4/4 bar, then cut/paste it to the last newly created bar).

### Prior art

_No response_

### Additional context

Similar issues potentially exist with respect to key and clef changes; the solution should seem consistent across all through cases.  Also, probably, with respect to double barlines etc.  Would be nice if append vs somehow did things that seemed intuitive for all of these cases.  But time signatures are the worst case right now because attempting to fiddle with them cause rebarring of the music that you almost certainly don't want.

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #18149 

Insert after selection now behaves differently, comparing to insert before
It doesn't move clefs, key and time signatures

`InsertMeasureOptions` already have `moveSignaturesClef` property, but code for Clefs were broken, it is fixed now.

(~~It doesn't handle barlines, as they are inconsitent also in "insert measure before selection" - "standard" barlines behave differently, than repeat barlines.~~)

Also barlines are correctly moved now.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

https://github.com/musescore/MuseScore/assets/1646034/786b297c-4570-4c37-b914-75e923abdff6

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/dom/edit.cpp [hunk 1]
```diff
@@ -4069,83 +4069,107 @@ MeasureBase* Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure,
...
+            if (measureInsert) {
                 for (size_t staffIdx = 0; staffIdx < score->nstaves(); ++staffIdx) {
-                    Measure* pm = measureInsert->prevMeasure();
+                    Measure* pm = newMeasure->prevMeasure();
                     if (pm) {
                         Segment* ps = pm->findSegment(SegmentType::Clef, tick);
                         if (ps && ps->enabled()) {
                             EngravingItem* pc = ps->element(staffIdx * VOICES);
-                            if (pc) {
+                            if (pc && !options.moveSignaturesClef) {
```
Comment (cbjeukendrup) [inline src/engraving/dom/edit.cpp]: It looks like we can move the `!options.moveSignaturesClef` a few `if` conditions up, to save some function calls.

File: src/engraving/dom/edit.cpp [hunk 2]
```diff
@@ -4069,83 +4069,107 @@ MeasureBase* Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure,
...
+                            localMeasure->undoChangeProperty(Pid::REPEAT_START, true);
+                            measureInsert->undoChangeProperty(Pid::REPEAT_START, false);
                         }
-                        if (ee) {
-                            undo(new RemoveElement(ee));
-                            if (s->empty()) {
-                                undoRemoveElement(s);
-                            }
+                    } else {
+                        if (pm->repeatEnd()) {
```
Comment (cbjeukendrup) [inline src/engraving/dom/edit.cpp]: Here you are using `pm` outside the `if (pm)` condition, so it might be null.

**Review by cbjeukendrup — COMMENTED** [review]
I had a couple of remarks, but to save you some time I fixed them myself. Could you please check if everything looks right? 

(It's the most convenient to review all my commits separately, and to use the "hide whitespace" setting.)

**Comment by sammik** [comment]
@cbjeukendrup 
Thank you, wery good changes, I had just one think to change, I posted review.

But if we doing this optimisations, may be it would be better also to have just one `staves` loop instead of three.

I post one other commit with proposed changes, please have a look.

**Comment by cbjeukendrup** [comment]
@sammik Actually I made the for-loops consciously as local as possible, so that those `if` conditions are evaluated only once instead of for every staff again... what do you think?

**Comment by sammik** [comment]
whops, 

> @sammik Actually I made the for-loops consciously as local as possible, so that those `if` conditions are evaluated only once instead of for every staff again... what do you think?

whops, sorry
@cbjeukendrup You are right, of course!

so, finaly, I changed only that one thing, I rewieved above and added two nullchecks

thanks for patience, please check it now

**Comment by cbjeukendrup** [comment]
(One more push, to rebase, and to fix a mistake I had made in the first refactor commit (that you had also fixed in your refactor commit, but I wanted to make sure that my commit was "self-contained"))

**Comment by cbjeukendrup** [comment]
I found one potential inconsistency:
- when you select the measure before an "end repeat" barline in the middle of the score, "after selection" will place the new measure before the end repeat barline
- when the _last_ measure of the score has an "end repeat" barline, and you do "after selection" on that last measure, the new measure will be placed _after_ the end repeat barline.

**Comment by sammik** [comment]
> I found one potential inconsistency:
> 
> * when you select the measure before an "end repeat" barline in the middle of the score, "after selection" will place the new measure before the end repeat barline
> * when the _last_ measure of the score has an "end repeat" barline, and you do "after selection" on that last measure, the new measure will be placed _after_ the end repeat barline.

Yes, thanks, Ill trake a look.

Should be "after last bar" same as "append to the and of the score"? (Both move repeat barline to the and of the score?)

**Comment by sammik** [comment]
@cbjeukendrup done

**Comment by cbjeukendrup** [comment]
@sammik Sorry, I found one more issue... namely, the same goes for non-repeat barlines. If an end barline is customised, "insert after" will put the new measure after the custom barline, and if any other barline is customised, the new measure is placed before it.

**Comment by cbjeukendrup** [comment]
Another one: if you select the title frame, then use Add > Measures > After selection, the newly inserted bar is placed before the initial key and time signatures. 
Scherm­afbeelding 2023-09-29 om 22 30 27

**Comment by cbjeukendrup** [comment]
> Should be "after last bar" same as "append to the and of the score"? (Both move repeat barline to the and of the score?)

I think what you did now is fine. This way, the user can achieve both options.

**Comment by sammik** [comment]
@cbjeukendrup all done :)

**Review by cbjeukendrup — APPROVED** [review]

**Pull request merged**